ocrd_utils/ocrd_utils/os.py

```diff
@@ -98,33 +100,45 @@ def list_all_resources(executable):
     systemdir = Path('/usr/local/share/ocrd-resources', executable)
     if systemdir.is_dir():
         candidates += systemdir.iterdir()
+    if moduled:
+        # recurse fully
+        for resource in itertree(Path(moduled)):
+            if resource.is_dir():
+                continue
+            if any(resource.match(pattern) for pattern in
+                   # Python distributions do not distinguish between
+                   # code and data; `is_resource()` only singles out
+                   # files over directories; but we want data files only
+                   # todo: more code and cache exclusion patterns!
+                   ['*.py', '*.py[cod]', '*~', 'ocrd-tool.json']):
+                continue
+            candidates.append(resource)
     # recurse once
     for parent in candidates:
         if parent.is_dir():
             candidates += parent.iterdir()
-    return [str(x) for x in candidates]
+    return sorted([str(x) for x in candidates])
 
 def get_processor_resource_types(executable, ocrd_tool=None):
     """
-    Determine whether a processor has resource parameters that represent
-    directories (``has_dirs``), files (``has_files``) or neither.
+    Determine what type of resource parameters a processor needs.
 
-    Returns a pair ``(has_dir, has_files)``
+    Return a list of MIME types (with the special value `*/*` to
+    designate that arbitrary files or directories are allowed).
```

Comment:
That is much better than `has_dirs, has_files`.

Reply:
It's been quite a chase to get here, but yes, I think this makes sense now.